aesd-char-driver/main.c

```diff
@@ -18,51 +18,150 @@
+        else{
+            retval=data_size;
+            *f_pos+=data_size;
+        }
+    }
+    else{
+        *f_pos=0;
+        retval=0;
+    }
+    mutex_unlock(&data->char_dev_mutex_lock);    
     return retval;
 }
 
-ssize_t aesd_write(struct file *filp, const char __user *buf, size_t count,
+ssize_t packet_buffer_size=0;
+char* packet_buffer=NULL;
+ssize_t aesd_write(struct file
+ *filp, const char __user *buf, size_t count,
                 loff_t *f_pos)
 {
-    ssize_t retval = -ENOMEM;
+    ssize_t retval=-ENOMEM;
+    int delimiter_index=0,ret;
+    struct aesd_dev *driver_data=filp->private_data; 
+    size_t packet_start=0,allocation_size=0;
+    char* kernel_buffer=NULL,*removed_entry;
+    struct aesd_buffer_entry aesd_buf;
+
     PDEBUG("write %zu bytes with offset %lld",count,*f_pos);
-    /**
-     * TODO: handle write
-     */
+    retval=mutex_lock_interruptible(&driver_data->char_dev_mutex_lock);
+    if(retval!=0){
+        PDEBUG("Error while locking mutex");
+        return retval;
+    }
+    kernel_buffer=kmalloc(count,GFP_KERNEL);
+    if(kernel_buffer!=NULL){
+        ret=copy_from_user(kernel_buffer,buf,count);
+        PDEBUG("Kernel buffer is:%s",kernel_buffer);
+        if(ret!=0){
+            retval=-EFAULT;
+            PDEBUG("Copying from user failed");
+            goto exit_mutex_unlock;
+        }
+       do{
+        delimiter_index=find_new_line(kernel_buffer,(count-packet_start));
+        PDEBUG("Delimiter index : %d",delimiter_index);
+        allocation_size=(delimiter_index==-1)?(count-packet_start):((delimiter_index-packet_start)+1);
+        PDEBUG("Allocation_size is:%ld",allocation_size);
+        if(packet_buffer_size==0){
+            packet_buffer=kmalloc(allocation_size,GFP_KERNEL);
+            if(packet_buffer==NULL){
+                PDEBUG("Mallocation failed");
+                goto exit_mutex_unlock;
+            }
+        }
+        if(packet_buffer_size>0){
+            packet_buffer=krealloc(packet_buffer,(allocation_size+packet_buffer_size),GFP_KERNEL);
+            if(packet_buffer==NULL){
+                PDEBUG("Reallocation failed");
+                goto exit_mutex_unlock;
+            }         
+        }
+        //Performing memcpy
+        memcpy((packet_buffer+packet_buffer_size),(kernel_buffer+packet_start),allocation_size);
+        PDEBUG("Packet buffer:%s",packet_buffer);
+        packet_buffer_size+=allocation_size;
+        PDEBUG("Packet buffer size:%ld",packet_buffer_size);
+        if(delimiter_index!=-1){
+            aesd_buf.buffptr=packet_buffer;
+            aesd_buf.size=packet_buffer_size;
+            removed_entry=aesd_circular_buffer_add_entry(&driver_data->c_buffer,&aesd_buf);
+            if(removed_entry!=NULL){
+                kfree(removed_entry);
+            }
+            packet_buffer_size=0;
+            packet_start=packet_start+delimiter_index+1;
+        }
+        retval=count;
+        }
+        while(delimiter_index!=-1);
```

Comment:
this leads to a single thread holding the lock until it completes by sending newline charcater